src/runtime/micro/micro_session.cc

```diff
@@ -0,0 +1,309 @@
+    // First, retrieve the string `last_error` points to.
+    std::uintptr_t last_err_data_addr;
+    low_level_device()->Read(last_err_offset, &last_err_data_addr, sizeof(std::uintptr_t));
+    dev_base_offset last_err_data_offset =
+        dev_addr(last_err_data_addr) - low_level_device()->base_addr();
+    // Then read the string from device to host and log it.
+    std::string last_error_str = ReadString(last_err_data_offset);
+    LOG(FATAL) << "error during micro function execution:\n"
+               << "  " << last_error_str;
+  }
+}
+
+BinaryInfo MicroSession::LoadBinary(std::string binary_path) {
+  SectionLocation text;
+  SectionLocation rodata;
+  SectionLocation data;
+  SectionLocation bss;
+
+  text.size = GetSectionSize(binary_path, kText);
+  rodata.size = GetSectionSize(binary_path, kRodata);
+  data.size = GetSectionSize(binary_path, kData);
+  bss.size = GetSectionSize(binary_path, kBss);
+
+  text.start = AllocateInSection(kText, text.size);
+  rodata.start = AllocateInSection(kRodata, rodata.size);
+  data.start = AllocateInSection(kData, data.size);
+  bss.start = AllocateInSection(kBss, bss.size);
+  CHECK(!text.start.is_null() && !rodata.start.is_null() && !data.start.is_null() &&
+        !bss.start.is_null()) << "not enough space to load module on device";
+  const dev_base_addr base_addr = low_level_device_->base_addr();
+  std::string relocated_bin = RelocateBinarySections(
+      binary_path,
+      text.start + base_addr,
+      rodata.start + base_addr,
+      data.start + base_addr,
+      bss.start + base_addr);
+  std::string text_contents = ReadSection(relocated_bin, kText);
+  std::string rodata_contents = ReadSection(relocated_bin, kRodata);
+  std::string data_contents = ReadSection(relocated_bin, kData);
+  std::string bss_contents = ReadSection(relocated_bin, kBss);
+  low_level_device_->Write(text.start, &text_contents[0], text.size);
+  low_level_device_->Write(rodata.start, &rodata_contents[0], rodata.size);
+  low_level_device_->Write(data.start, &data_contents[0], data.size);
+  low_level_device_->Write(bss.start, &bss_contents[0], bss.size);
+  SymbolMap symbol_map {relocated_bin, base_addr};
+  return BinaryInfo{
+      .text = text,
+      .rodata = rodata,
+      .data = data,
+      .bss = bss,
+      .symbol_map = symbol_map,
+  };
+}
+
+void MicroSession::SetInitBinaryPath(std::string path) {
+  init_binary_path_ = path;
+}
+
+dev_addr MicroSession::EncoderWrite(TargetDataLayoutEncoder* encoder, UTVMArgs* args) {
+  auto utvm_args_slot = encoder->Alloc<UTVMArgs>();
+
+  const int* type_codes = args->type_codes;
+  int num_args = args->num_args;
+
+  auto tvm_vals_slot = encoder->Alloc<TVMValue*>(num_args);
+  auto type_codes_slot = encoder->Alloc<const int>(num_args);
+
+  for (int i = 0; i < num_args; i++) {
+    switch (type_codes[i]) {
+      case kNDArrayContainer:
+      case kArrayHandle: {
+        TVMArray* arr_handle = reinterpret_cast<TVMArray*>(args->values[i].v_handle);
```

Comment:
don't do an explicit cast. directly do TVMArray* arr_handle = args->values[i]; this will invoke the conversion function. To help us to be future compatible.